src/hotspot/share/nmt/memLogRecorder.cpp

```diff
@@ -0,0 +1,826 @@
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+// record pattern of allocations of memory calls:
+//
+// ./build/macosx-aarch64-server-release/xcode/build/jdk/bin/java -XX:+UnlockDiagnosticVMOptions -XX:NMTRecordMemoryAllocations=0x7FFFFFFF -jar build/macosx-aarch64-server-release/images/jdk/demo/jfc/J2Ddemo/J2Ddemo.jar
+//
+// OR record pattern of allocations of virtual memory calls:
+//
+// ./build/macosx-aarch64-server-release/xcode/build/jdk/bin/java -XX:+UnlockDiagnosticVMOptions -XX:NMTRecordVirtualMemoryAllocations=0x7FFFFFFF -jar build/macosx-aarch64-server-release/images/jdk/demo/jfc/J2Ddemo/J2Ddemo.jar
+//
+// this will produce 3 files:
+//
+// #1 hs_nmt_pid22770_allocs_record.log (is the chronological record of the the desired operations)
+// OR
+// #1 hs_nmt_pid22918_virtual_allocs_record.log (is the chronological record of the desired operations)
+// #2 hs_nmt_pid22770_info_record.log (is the record of default NMT memory overhead and the NMT state)
+// #3 hs_nmt_pid22770_threads_record.log (is the record of thread names that can be retrieved later when processing)
+//
+// then to actually run the benchmark:
+//
+// ./build/macosx-aarch64-server-release/xcode/build/jdk/bin/java -XX:+UnlockDiagnosticVMOptions -XX:NMTBenchmarkRecordedPID=43100 -XX:NMTBenchmarkRecordedLoops=10
+
+#include "precompiled.hpp"
+#include "jvm.h"
+#include "nmt/nmtCommon.hpp"
+#include "nmt/mallocHeader.hpp"
+#include "nmt/mallocHeader.inline.hpp"
+#include "nmt/memLogRecorder.hpp"
+#include "nmt/memReporter.hpp"
+#include "nmt/memTracker.hpp"
+#include "runtime/arguments.hpp"
+#include "runtime/globals.hpp"
+#include "runtime/os.hpp"
+#include "runtime/thread.hpp"
+#include "utilities/debug.hpp"
+#include "utilities/nativeCallStack.hpp"
+#include "utilities/permitForbiddenFunctions.hpp"
+#include "utilities/vmError.hpp"
+
+#if defined(LINUX)
+#include <malloc.h>
+#elif defined(__APPLE__)
+#include <malloc/malloc.h>
+#endif
+
+#if defined(LINUX) || defined(__APPLE__)
+#include <pthread.h>
+#include <string.h>
+#include <sys/mman.h>
+#endif
+
+NMT_MemoryLogRecorder NMT_MemoryLogRecorder::_recorder;
+NMT_VirtualMemoryLogRecorder NMT_VirtualMemoryLogRecorder::_recorder;
+
+void NMT_LogRecorder::initialize(intx memoryCount, intx virtualMemoryCount) {
+  NMT_MemoryLogRecorder::initialize(NMTRecordMemoryAllocations);
+  NMT_VirtualMemoryLogRecorder::initialize(NMTRecordVirtualMemoryAllocations);
+}
+
+void NMT_LogRecorder::finish() {
+  NMT_MemoryLogRecorder::instance()->finish();
+  NMT_VirtualMemoryLogRecorder::instance()->finish();
+}
+
+void NMT_LogRecorder::replay(const char* path, const int pid) {
+  NMT_MemoryLogRecorder::instance()->replay(path, pid);
+  NMT_VirtualMemoryLogRecorder::instance()->replay(path, pid);
+}
+
+void NMT_LogRecorder::init() {
+#if defined(LINUX) || defined(__APPLE__)
+  pthread_mutex_t _mutex;
+#elif defined(WINDOWS)
+ // ???
+#endif
+  pthread_mutex_init(&_mutex, NULL);
```

Comment:
You need to use existing synchronization API here. Either Mutex or PlatformMutex depending on initialization constraints; maybe a Semaphore if the others can't work.